File: MatlabCode/Repo_CompIntroQP_MATLAB/ChapterWF/Exercise_1_6_4_a.m
% This script estimates expectation values for r
% and p for the ground state of Hydrogen.
% It does so using SI units.
%
% Inputs:
%   L       - The extension of the radial grid 
%   N       - The number of grid points

% Numerical grid parameters
L = input('Please provide the extension of your grid (in metres): ');
N = input('Please provide the number of grid points: ');

% Relevant constants in SI units
a0 = 5.292e-11;                 % The Bohr radius
hbar = 1.055e-34;               % The reduced Planck constant 

% Provide wave function
PsiFunk = @(r) 2/a0^(3/2)*r.*exp(-r/a0);

% Set up grid
r = linspace(0, L, N);
Psi = PsiFunk(r);               % Vector with function values
h = L/(N-1);                    % Increment 

% Calculate expectation value for r
MeanR = trapz(r, r.*abs(Psi).^2);                 % Mean position
% Write result to screen
disp(['Mean distace: ', num2str(MeanR), ' m'])

% Set up vector with Psi'(r)
PsiDeriv = zeros(1, N);             % Allocate
% End points (Assuming Psi(0) = Psi(R) = 0)
PsiDeriv(1) =  Psi(2)/(2*h);    
PsiDeriv(N) = -Psi(N-1)/(2*h);
% Estimate the derivative with the midpoint rule
for n = 2:(N-1)
  PsiDeriv(n) = (Psi(n+1)-Psi(n-1))/(2*h);
end

% Calculate expectation value for p
MeanP = -1i*hbar*trapz(r, conj(Psi).*PsiDeriv);   % Mean momentum
% Write result to screen
disp(['Mean momentum (real part): ', num2str(real(MeanP)), ' kg*m/s'])

File: PythonCode/ChapterWF/Exercise_1_6_4_a.py
""" 
This script estimates expectation values for r
and p for the ground state of Hydrogen.
It does so using SI units.

Inputs:
   L       - The extension of the radial grid 
   N       - The number of grid points

The input function is used to provide the input parameters from screen.
"""

# Import the numpy library
import numpy as np

# Numerical grid parameters
L = float(input('Please provide the extension of your grid (in metres): '))
N = int(input('Please provide the number of grid points: '))

# Relevant constants in SI units
a0 = 5.292e-11                # The Bohr radius
hbar = 1.055e-34              # The reduced Planck constant 

# Normalzied wave function
def  PsiFunk(r):
  return 2/a0**(3/2)*r*np.exp(-r/a0)


# Set up grid
r = np.linspace(0, L, N)
Psi = PsiFunk(r)              # Vector with function values
h = L/(N-1)                   # Increment 

# Calculate expectation value for r
MeanX = sum(r*np.abs(Psi)**2)*h                # Mean position
print(f'Mean position: {MeanX:.4e} m')

# Set up vector with Psi'(r)
PsiDeriv = np.zeros(N, dtype=complex)             # Allocate and declare
# End points (assume Psi = 0 outside the interval)
PsiDeriv[0] =  Psi[1]/(2*h)
PsiDeriv[N-1] = -Psi[N-1]/(2*h)
# Estimate the derivative with the midpoint rule
for n in range(1,N-1):
  PsiDeriv[n] = (Psi[n+1]-Psi[n-1])/(2*h)


# Calculate expectation value for p
MeanP = -1j*hbar*sum(np.conj(Psi)*PsiDeriv)*h   
print(f'Mean momentum: {np.real(MeanP):.4e} kg m/s')
print(f'Imaginary part: {np.imag(MeanP):.4e} kg m/s')